### PR TITLE
feat(raft): assign priorities for nodes for priority election

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Priorities for nodes are calculated and assigned when the partition distribution is calculated. When priority election is enabled, this priority will be used. Otherwise the priority has no effect.

This PR is in preparation to #7285 

## Related issues

Related #7285 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java [hunk 1]
```diff
@@ -250,11 +251,51 @@ public String toString() {
...
+    } else {
+      int nextPriority = replicationFactor - 1;
+      for (final MemberId member : membersForPartition) {
+        if (!member.equals(primary)) {
+          priority.put(member, nextPriority);
+          nextPriority--;
+        }
+      }
+    }
+    return priority;
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java:296]: I think we can extract that into a separate class and write an unit test for it.

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java [hunk 2]
```diff
@@ -250,11 +251,51 @@ public String toString() {
       for (int j = 0; j < count; j++) {
         membersForPartition.add(sorted.get((i + j) % length));
       }
-      metadata.add(new PartitionMetadata(partitionId, membersForPartition));
+      final var primary = sorted.get(i % length);
+      final var priority =
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java:255]:
```suggestion
      final var priorities =
```

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @deepthidevaki :+1: 

I would like to see some unit tests, to verify that the priorities are set (calculated) as expected for some example values. 

Example test cases:
 *  homogeneous cluster (3,3,3) 
 * more partitions then nodes
 *  more nodes then partitions

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java (hunk 1 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java:296]: :+1:

**Comment by deepthidevaki** [comment]
@Zelldon Thanks for you review. I have extracted partition distribution to a class and added unit tests for it. I feel it could be structured in a better way, but we can revisit it when refactoring the bootstrap process.

File: atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        metadata -> {
+          final int partitionId = metadata.id().id();
+          metadata
+              .members()
+              .forEach(
+                  member -> {
+                    final int nodeId = Integer.parseInt(member.id());
+                    assertThat(metadata.getPriority(member))
+                        .describedAs(
+                            "Priority calculated wrong for nodeId %d, partition %d. Observed metadata is %s",
```
Comment (Zelldon) [inline atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java:58]: Might make sense to print also the expected :)

File: atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        Arguments.of(
+            3, // clusterSize
+            3, // partitionCount
+            3, // replicationFactor
+            // Expected partitionDistribution matrix row = nodes, column = partitions
+            new int[][] {
+              {3, 1, 2},
+              {2, 3, 1},
+              {1, 2, 3},
+            }),
```
Comment (Zelldon) [inline atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java:76]: :+1:

File: atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java [hunk 5]
```diff
@@ -179,7 +178,11 @@ public String toString() {
...
     // First member of a PartitionMetadata is the bootstrap leader
-    metadata = buildPartitions();
+    final var members =
+        config.getMembers().stream().map(MemberId::from).collect(Collectors.toSet());
+    metadata =
+        new PartitionDistributor()
+            .generatePartitionDistribution(members, sortedPartitionIds, replicationFactor);
     // +------------------+----+----+----+---+
     // | Partition \ Node | 0  | 1  | 2  | 3 |
     // +------------------+----+----+----+---+
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java:188]: Do we want to update this comment?

File: atomix/cluster/src/main/java/io/atomix/raft/partition/PartitionDistributor.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      final int replicationFactor) {
+    final Map<MemberId, Integer> priority = new HashMap<>();
+    final int lowestPriority = 1;
+
+    priority.put(primary, replicationFactor);
+    // To ensure that secondary priorities are distributed evenly, we alternate the nodes for which
+    // second priority is assigned. Example, clusterSize = 3 partitionCount = 12. Node 0 has highest
+    // priority (=3) for partition 1,4,7 and 9. For partition 1 and 7, node 1 gets priority 2. For
+    // partition 4 and 9, node 2 gets priority 2. This is done so that if node 0 dies, the
+    // leadership is evenly distributed on the rest of the followers.
```
Comment (Zelldon) [inline atomix/cluster/src/main/java/io/atomix/raft/partition/PartitionDistributor.java:78]:
```
Distribution:
P\N|	N 0|	N 1|	N 2
P 1|	L  |	F  |	F  
P 2|	F  |	L  |	F  
P 3|	F  |	F  |	L  
P 4|	L  |	F  |	F  
P 5|	F  |	L  |	F  
P 6|	F  |	F  |	L  
P 7|	L  |	F  |	F  
P 8|	F  |	L  |	F  
P 9|	F  |	F  |	L  
P 10|	L  |	F  |	F  
P 11|	F  |	L  |	F  
P 12|	F  |	F  |	L  
```

I think node 0 has highest prio for 10 and not 9

**Review by ChrisKujawa — APPROVED** [review]
Awesome thanks for extracting it, such that we can have some tests for it :100: 

I had some smaller hints for the tests, but not blockers.

>  I feel it could be structured in a better way, but we can revisit it when refactoring the bootstrap process.

You mean the class? Yeah we can probably iterate over it again :)

Good work :+1:

File: atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java (hunk 3 above)
Comment (deepthidevaki) [reply atomix/cluster/src/test/java/io/atomix/raft/partition/PartitionDistributorTest.java:58]: That would be printed by default. This is how it looks like:
```
org.opentest4j.AssertionFailedError: [Priority calculated wrong for nodeId 1, partition 3. Observed metadata is PartitionMetadata{id=PartitionId{id=3, group=test}, members=[2, 0, 1], priority={1=1, 2=3, 0=2}, targetPriority=3}] 
expected: 0
 but was: 1
Expected :0
Actual   :1
```

**Comment by deepthidevaki** [comment]
> > I feel it could be structured in a better way, but we can revisit it when refactoring the bootstrap process.
> 
> You mean the class? Yeah we can probably iterate over it again :)

Yes. `PartitionDistributor` is still tighly coupled with `RaftPartitionGroup`.

**Comment by deepthidevaki** [comment]
Flaky AdvertisedAddressTest

bors r+

**Pull request merged**